### ISSUE TITLE
Incorrect processing of XYWH boxes in `scale_boxes`

### ISSUE THREAD

**Issue by anmolojha** [ultralytics/ultralytics#20895]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Other

### Bug

`scale_boxes` function's docstring says that it supports both `xyxy` and `xywh` box formats, but its logic for `xywh` format is incorrect.

`scale_boxes` calls `clip_boxes` towards the end:

https://github.com/ultralytics/ultralytics/blob/7564930f1ad841ab9972712c5508c47c89480e54/ultralytics/utils/ops.py#L140

but `clip_boxes` does not handles `xywh` format. Therefore, `scale_boxes` can return boxes like `[110.,  30., 180., 180.]` (see example), which run out of the image.

### Environment

```
Ultralytics 8.3.146 🚀 Python-3.11.12 torch-2.6.0+cu124 CPU (Intel Xeon 2.20GHz)
Setup complete ✅ (2 CPUs, 12.7 GB RAM, 41.4/107.7 GB disk)

OS                  Linux-6.1.123+-x86_64-with-glibc2.35
Environment         Colab
Python              3.11.12
Install             pip
Path                /usr/local/lib/python3.11/dist-packages/ultralytics
RAM                 12.67 GB
Disk                41.4/107.7 GB
CPU                 Intel Xeon 2.20GHz
CPU count           2
GPU                 None
GPU count           None
CUDA                None

numpy               ✅ 2.0.2>=1.23.0
matplotlib          ✅ 3.10.0>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 11.2.1>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.15.3>=1.4.1
torch               ✅ 2.6.0+cu124>=1.8.0
torch               ✅ 2.6.0+cu124!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.21.0+cu124>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 5.9.5
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.2>=1.1.4
ultralytics-thop    ✅ 2.0.14>=2.0.0
{'OS': 'Linux-6.1.123+-x86_64-with-glibc2.35',
 'Environment': 'Colab',
 'Python': '3.11.12',
 'Install': 'pip',
 'Path': '/usr/local/lib/python3.11/dist-packages/ultralytics',
 'RAM': '12.67 GB',
 'Disk': '41.4/107.7 GB',
 'CPU': 'Intel Xeon 2.20GHz',
 'CPU count': 2,
 'GPU': None,
 'GPU count': None,
 'CUDA': None,
 'Package Info': {'numpy': '✅ 2.0.2>=1.23.0',
  'matplotlib': '✅ 3.10.0>=3.3.0',
  'opencv-python': '✅ 4.11.0.86>=4.6.0',
  'pillow': '✅ 11.2.1>=7.1.2',
  'pyyaml': '✅ 6.0.2>=5.3.1',
  'requests': '✅ 2.32.3>=2.23.0',
  'scipy': '✅ 1.15.3>=1.4.1',
  'torch': '✅ 2.6.0+cu124!=2.4.0,>=1.8.0; sys_platform == "win32"',
  'torchvision': '✅ 0.21.0+cu124>=0.9.0',
  'tqdm': '✅ 4.67.1>=4.64.0',
  'psutil': '✅ 5.9.5',
  'py-cpuinfo': '✅ 9.0.0',
  'pandas': '✅ 2.2.2>=1.1.4',
  'ultralytics-thop': '✅ 2.0.14>=2.0.0'}}
```

### Minimal Reproducible Example

```python
import torch
from ultralytics.utils.ops import scale_boxes, xyxy2xywh, xywh2xyxy

img1_shape = (320, 320)  # shape of processed image
img0_shape = (480, 640)  # shape of original image

boxes = torch.tensor([[10, 10, 100, 100]], dtype=torch.float32)  # xyxy

scaled_xyxy = scale_boxes(img1_shape=img1_shape, boxes=boxes.clone(), img0_shape=img0_shape, xywh=False)
print("Scaled XYXY:", scaled_xyxy)

boxes_xywh = xyxy2xywh(boxes)
scaled_xywh = scale_boxes(img1_shape=img1_shape, boxes=boxes_xywh.clone(), img0_shape=img0_shape, xywh=True)
print("Scaled XYWH:", scaled_xywh)
print("Scaled XYWH -> XYXY:", xywh2xyxy(scaled_xywh))
```

Output:
```
Scaled XYXY: tensor([[ 20.,   0., 200., 120.]])
Scaled XYWH: tensor([[110.,  30., 180., 180.]])
Scaled XYWH -> XYXY: tensor([[ 20., -60., 200., 120.]])
```

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @anmolojha, thank you for bringing this up and for providing such a detailed report and minimal reproducible example (MRE) 🚀! This helps us a lot in diagnosing and addressing the issue.

Our team recommends checking out the [Docs](https://docs.ultralytics.com/) for more context on box formats and utility functions. If you haven’t already, please ensure you’re running the latest `ultralytics` package, as many issues are resolved with updates. You can upgrade with:

```bash
pip install -U ultralytics
```

If you’re running into this bug in a particular environment, YOLO can be executed in a variety of up-to-date and verified setups (with all dependencies, including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

For real-time help and discussion, consider joining the Ultralytics community on [Discord](https://discord.com/invite/ultralytics) 🎧, [Discourse](https://community.ultralytics.com/), or our [Subreddit](https://reddit.com/r/Ultralytics).

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across platforms.

---

This is an automated response to help you get started quickly. An Ultralytics engineer will review your issue and assist you soon. Thank you again for your contribution and for offering to submit a PR! 🙏

**Comment by glenn-jocher** [comment]
Thank you for the detailed bug report! You're absolutely correct - the issue is that `clip_boxes` assumes `xyxy` format but `scale_boxes` calls it regardless of the `xywh` parameter. When `xywh=True`, the function incorrectly clips center coordinates and dimensions as if they were corner coordinates.

Your MRE clearly demonstrates the problem where the scaled XYWH boxes result in invalid coordinates when converted back to XYXY format. 

Since you're willing to submit a PR, the fix would need to handle the `xywh` format properly before clipping - either by converting to `xyxy` format before `clip_boxes`, adding `xywh` support to `clip_boxes`, or implementing separate clipping logic for `xywh` format within `scale_boxes`. We'd appreciate your contribution to resolve this!

**Comment by anmolojha** [comment]
Hi,
> either by converting to xyxy format before clip_boxes, adding xywh support to clip_boxes, or implementing separate clipping logic for xywh format within scale_boxes

Thanks for outlining the options. Here's my understanding and thoughts on each approach:

---

### 1. Converting to xyxy format before clip_boxes
Small change but changing format (`xywh2xyxy` and `xyxy2xywh`) requires making a new tensor.
https://github.com/ultralytics/ultralytics/blob/5921074a34c19faa12e5303d313577d5dd053ce5/ultralytics/utils/ops.py#L432

### 2. Adding xywh support to clip_boxes
Earlier, I was leaning towards this. I thought that `clip_boxes` is a general utility function, and should support all formats, but there are a lot of box formats (`xywh`, `xyxy`, `ltwh` etc). Allowing only `xyxy` and `xywh` by adding an argument `xywh=True` seems a little ad-hoc. 

I think a better way to implement this would be to have a `box_format` argument like:
```python
box_format= "xywh" | "xyxy" | "ltwh" ...
```
But I think that's a larger discussion and should be thought of at the codebase level. *(Maybe use `Boxes` class everywhere to store data and format together?)*

### 3. Implementing separate clipping logic for xywh format within scale_boxes
Not very keen on this. `clip_boxes` already exists to handle clipping logic, so duplicating that logic elsewhere (even partially) doesn't seem ideal.

---

Overall, I'm currently leaning toward Option 1 due to its simplicity and consistency with existing code. For example, the following snippet follows the same approach:
https://github.com/ultralytics/ultralytics/blob/5921074a34c19faa12e5303d313577d5dd053ce5/ultralytics/utils/callbacks/comet.py#L166-L170

Let me know your thoughts. Will implement changes accordingly.

Thanks!

**Comment by glenn-jocher** [comment]
Your analysis is excellent and I completely agree with Option 1. The format conversion approach maintains consistency with existing patterns in the codebase, like the [comet callback example you referenced](https://docs.ultralytics.com/reference/utils/callbacks/comet/#ultralytics.utils.callbacks.comet._scale_bounding_box_to_original_image_shape), and keeps `clip_boxes` focused on its current responsibility.

You're right that expanding `clip_boxes` to handle multiple formats would require broader architectural discussion, and duplicating clipping logic in `scale_boxes` isn't ideal. The temporary tensor creation for format conversion is a reasonable trade-off for the cleaner separation of concerns.

Looking forward to your PR implementing Option 1!
